src/migrations/rebase.rs

```diff
@@ -0,0 +1,286 @@
+use crate::migrations::create::{MigrationKey, MigrationToText};
+use crate::migrations::db_migration::{DBMigration, read_all};
+use crate::migrations::dev_mode::rebase_to_schema;
+use crate::migrations::extract::DatabaseMigration;
+use crate::migrations::grammar::parse_migration;
+use crate::migrations::migrate::apply_migrations_inner;
+use crate::migrations::migration::MigrationFile;
+use crate::print;
+
+#[derive(PartialEq)]
+enum RebaseMigrationKind {
+    Base,
+    Source,
+    Target,
+}
+
+struct RebaseMigration<'a> {
+    key: MigrationKey,
+    migration: &'a DBMigration,
+    parent: Option<&'a str>,
+    kind: RebaseMigrationKind
+}
+
+impl MigrationToText for RebaseMigration<'_> {
+    type StatementsIter<'a> = std::iter::Once<&'a String> where Self: 'a;
+
+    fn key(&self) -> &MigrationKey {
+        &self.key
+    }
+
+    fn parent(&self) -> anyhow::Result<&str> {
+        if let Some(parent) = self.parent {
+            return Ok(parent)
+        }
+
+        if self.migration.parent_names.is_empty() {
+            return Ok("initial");
+        }
+
+        Ok(&self.migration.parent_names[0])
+    }
+
+    fn id(&self) -> anyhow::Result<&str> {
+        Ok(&self.migration.name)
+    }
+
+    fn statements<'a>(&'a self) -> Self::StatementsIter<'a> {
+        std::iter::once(&self.migration.script)
+    }
+}
+
+pub struct RebaseMigrations {
+    /// initial..base : the commonly shared migrations between both 'source' and 'target'
+    base_migrations: IndexMap<String, DBMigration>,
+    /// base..source : the migrations that are unique to 'source'
+    source_migrations: IndexMap<String, DBMigration>,
+    /// base..target : the migrations that are unique to 'target'
+    target_migrations: IndexMap<String, DBMigration>,
+}
+
+impl RebaseMigrations {
+    pub fn print_status(&self) {
+        let last_common = self.base_migrations.last().map(|v| v.0.as_str()).unwrap_or("initial").green();
+
+        let format_migration_on_length = |c: usize| {
+             if c > 1 {"migrations"} else {"migration"}
+        };
+
+        eprintln!("Last common migration is {}", last_common);
+        eprintln!(
+            "Since then, there are:\n- {} new {} on the target branch,\n- {} {} to rebase",
+            self.target_migrations.len().to_string().green(), format_migration_on_length(self.target_migrations.len()),
+            self.source_migrations.len().to_string().green(), format_migration_on_length(self.source_migrations.len())
+        );
+    }
+
+    fn flatten(&self) -> anyhow::Result<Vec<RebaseMigration>> {
+        let mut result = Vec::new();
+        let mut key_index = 0;
+
+        let mut next_key = || -> MigrationKey {
+            key_index += 1;
+            MigrationKey::Index(key_index)
+        };
+
+        for (_, migration) in &self.base_migrations {
+            result.push(RebaseMigration {
+                key: next_key(),
+                parent: None,
```

Comment:
This is not correct, because you are iterating over all `base_migrations`.

```suggestion
                parent: result.last().map(|v| v.migration.name.as_str()),
```

Reply:
If `parent` is `None` on `RebaseMigration`, the `DBMigration`s' `parent_names` field is used, https://github.com/edgedb/edgedb-cli/blob/master/src/migrations/branch.rs#L32-L42

Reply:
which yields correct behavior for `base_migrations`

Reply:
But don't we iterate over all migrations, so we could have all `.parent`s set correctly, even for `base_migrations`?

It seems like unnecessary complication, but you are right, it does fallback correctly.

Reply:
> But don't we iterate over all migrations

In `get_diverging_migrations`?